src/extensions/common/useFetchCache.ts

```diff
@@ -0,0 +1,103 @@
+import objectHash from 'object-hash'
+import { useEffect, useMemo, useReducer, useRef } from "react"
+
+type FetchCache<QueryType> = {
+    get: (query: QueryType) => any | undefined
+}
+
+type FetchCacheState = {
+    data: {[key: string]: any}
+    activeFetches: {[key: string]: boolean}
+}
+
+const initialFetchCacheState = {
+    data: {},
+    activeFetches: {}
+}
+
+type StartFetchAction = {
+    type: 'startFetch',
+    queryHash: string
+}
+
+type SetDataAction = {
+    type: 'setData',
+    queryHash: string,
+    data: any
+}
+
+type FetchCacheAction = StartFetchAction | SetDataAction
+
+const fetchCacheReducer = (state: FetchCacheState, action: FetchCacheAction): FetchCacheState => {
+    switch(action.type) {
+        case 'startFetch': {
+            return {
+                ...state,
+                activeFetches: {
+                    ...state.activeFetches,
+                    [action.queryHash]: true
+                }
+            }
+        }
+        case 'setData': {
+            return {
+                ...state,
+                data: {
+                    ...state.data,
+                    [action.queryHash]: action.data
+                },
+                activeFetches: {
+                    ...state.activeFetches,
+                    [action.queryHash]: false
+                }
+            }
+        }
+        default: {
+            throw Error('Unexpected action in fetchCacheReducer')
+        }
+    }
+}
+
+const queryHash = <QueryType>(query: QueryType) => {
+    return objectHash(query)
+}
+
+const useFetchCache = <QueryType>(fetchFunction: (query: QueryType) => Promise<any>): FetchCache<QueryType> => {
+    const [state, dispatch] = useReducer(fetchCacheReducer, initialFetchCacheState)
+    const queriesToFetch = useRef<{[key: string]: QueryType}>({})
+    const get = useMemo(() => ((query: QueryType) => {
+        const h = queryHash(query)
+        const v = state.data[h]
+        if (v === undefined) {
+            queriesToFetch.current[h] = query
+        }
+        return v
+    }), [state.data])
+    const fetch = useMemo(() => ((query: QueryType) => {
```

Comment:
The usage distinction between `get` and `fetch` is not immediately obvious to me. I think the first one is retrieving the (cached) value and the other is requesting the (not yet cached) value? Do I have that right?